README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# swissexchangedata

<!-- badges: start -->

[![R-CMD-check](https://github.com/akhapwoyaco/swissexchangedata/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/akhapwoyaco/swissexchangedata/actions/workflows/R-CMD-check.yaml)
[![Codecov test
coverage](https://codecov.io/gh/akhapwoyaco/swissexchangedata/branch/main/graph/badge.svg)](https://app.codecov.io/gh/akhapwoyaco/swissexchangedata?branch=main)
<!-- badges: end -->

## Overview

The `swissexchangedata` package, built up for **educational purpose** to
learn and sharpen skills in web-scrapping of data from online sources,
data wrangling and package development processes.

swisswxchangedata facilitates ease of access of the
`Swiss Stock Exchange Market Data News` Via [The SIX Group Newsboard
Website](https://www.six-group.com/en/products-services/the-swiss-stock-exchange/market-data/news-tools/newsboard.html#/)
via the function: `newsboard_data`.

## Installation

You can install the development version of `swissexchangedata` like so:

``` r
install.packages("devtools") # if you have not installed "devtools" package
devtools::install_github("akhapwoyaco/swissexchangedata")
```

## Usage

This is a basic example which shows you how to solve a common problem:

``` r
library(swissexchangedata)
## basic example code
```

Get the data for 9th October 2023 up-to-date:  

``` r
data_10thOct_2024 <- newsboard_data(firstDate = '2023-10-10', lastDate = Sys.Date())
#> [1] "Page 0"
#> [1] "Page 0 has 4 observations."
#> [1] "Page 1"
#> [1] "Page 1 has  observations."
```

``` r
head(data_10thOct_2024)
#>   messageNo         isin valorSymbol                                      title
#> 1    208873 CH0344958688         E18    Mistrade Decision in E18 / CH0344958688
#> 2    208872 CH1272334934      KGIHJB Mistrade Decision in KGIHJB / CH1272334934
#> 3    208871 CH1240058615      KGIMJB Mistrade Decision in KGIMJB / CH1240058615
#> 4    208870 IE00BDR5H412      GENDES Mistrade Decision in GENDES / IE00BDR5H412
#>   messageType broadcastDateTime                      security
#> 1    Mistrade      2.023101e+13               0.50 EIDG 18-32
#> 2    Mistrade      2.023101e+13             KGIHJB JB P 12/23
#> 3    Mistrade      2.023101e+13             KGIMJB JB C 12/23
#> 4    Mistrade      2.023101e+13 UBSETF GLOBAL GENDER hCHF ACC
#>                    tradingSegment priority markets products currency
#> 1 Bonds - CHF Swiss Confederation   Normal    XSWX       BO      CHF
#> 2             Structured Products   Normal    XQMH       DE      CHF
#> 3             Structured Products   Normal    XQMH       DE      CHF
#> 4                             ETF   Normal    XSWX       FU      CHF
#>                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      newsText
#> 1                <div><p>In accordance with the rules of SIX Swiss Exchange, the following trade in <strong>'0.50 EIDG 18-32'</strong> has been declared a mistrade and has therefore been cancelled: </p><table><thead><tr><th align="center" style="width: 75px">Trade Date</th><th align="center" style="width: 75px">Time</th><th align="center" style="width: 45px">Cur</th><th align="left" style="width: 75px">Size</th><th align="left" style="width: 75px">Price</th><th align="left" style="width: 75px">Trade Type</th><th align="left" style="width: 75px">Book Type</th></tr></thead><tbody><tr><td align="center">11.10.2023</td><td align="center">11:21:55</td><td align="center">CHF</td><td align="left">10'000</td><td align="left">94.0000</td><td align="left">OnExchange</td><td align="left">Clob</td></tr></tbody></table><p>Please find further information concerning mistrades in Directive 4: Market Control on our website.</p><p>Regards,<br/>Exchange Operations, SIX Swiss Exchange</p></div>
#> 2           <div><p>In accordance with the rules of SIX Swiss Exchange, the following trade in <strong>'KGIHJB JB P 12/23'</strong> has been declared a mistrade and has therefore been cancelled: </p><table><thead><tr><th align="center" style="width: 75px">Trade Date</th><th align="center" style="width: 75px">Time</th><th align="center" style="width: 45px">Cur</th><th align="left" style="width: 75px">Size</th><th align="left" style="width: 75px">Price</th><th align="left" style="width: 75px">Trade Type</th><th align="left" style="width: 75px">Book Type</th></tr></thead><tbody><tr><td align="center">10.10.2023</td><td align="center">16:40:20</td><td align="center">CHF</td><td align="left">5'000</td><td align="left">0.7600</td><td align="left">OnExchange</td><td align="left">QuoteBook</td></tr></tbody></table><p>Please find further information concerning mistrades in Directive 4: Market Control on our website.</p><p>Regards,<br/>Exchange Operations, SIX Swiss Exchange</p></div>
#> 3           <div><p>In accordance with the rules of SIX Swiss Exchange, the following trade in <strong>'KGIMJB JB C 12/23'</strong> has been declared a mistrade and has therefore been cancelled: </p><table><thead><tr><th align="center" style="width: 75px">Trade Date</th><th align="center" style="width: 75px">Time</th><th align="center" style="width: 45px">Cur</th><th align="left" style="width: 75px">Size</th><th align="left" style="width: 75px">Price</th><th align="left" style="width: 75px">Trade Type</th><th align="left" style="width: 75px">Book Type</th></tr></thead><tbody><tr><td align="center">10.10.2023</td><td align="center">16:40:21</td><td align="center">CHF</td><td align="left">5'000</td><td align="left">0.6900</td><td align="left">OnExchange</td><td align="left">QuoteBook</td></tr></tbody></table><p>Please find further information concerning mistrades in Directive 4: Market Control on our website.</p><p>Regards,<br/>Exchange Operations, SIX Swiss Exchange</p></div>
#> 4 <div><p>In accordance with the rules of SIX Swiss Exchange, the following trade in <strong>'UBSETF GLOBAL GENDER hCHF ACC'</strong> has been declared a mistrade and has therefore been cancelled: </p><table><thead><tr><th align="center" style="width: 75px">Trade Date</th><th align="center" style="width: 75px">Time</th><th align="center" style="width: 45px">Cur</th><th align="left" style="width: 75px">Size</th><th align="left" style="width: 75px">Price</th><th align="left" style="width: 75px">Trade Type</th><th align="left" style="width: 75px">Book Type</th></tr></thead><tbody><tr><td align="center">10.10.2023</td><td align="center">11:35:19</td><td align="center">CHF</td><td align="left">41</td><td align="left">14.4180</td><td align="left">OnExchange</td><td align="left">QuoteBook</td></tr></tbody></table><p>Please find further information concerning mistrades in Directive 4: Market Control on our website.</p><p>Regards,<br/>Exchange Operations, SIX Swiss Exchange</p></div>
#>   newsTypeCode
#> 1           MI
#> 2           MI
#> 3           MI
#> 4           MI
```

## Getting help

If you encounter a clear bug, please file an issue with a minimal
reproducible example on GitHub.

## Code of conduct

Please note that this project is released with a Contributor Code of
Conduct. By participating in this project you agree to abide by its
terms.


NEWS.md
# swissexchangedata 0.1.0

* Initial CRAN submission.


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/testing-design.html#sec-tests-files-overview
# * https://testthat.r-lib.org/articles/special-files.html

library(testthat)
library(swissexchangedata)

test_check("swissexchangedata")


tests/testthat/test-output.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/testing-design.html#sec-tests-files-overview
# * https://testthat.r-lib.org/articles/special-files.html

# library(plyr)
# library(jsonlite)
# library(testthat)
# library(swissexchangedata)

# test that the output from newsboard_data() returns a dataframe class object
test_that("Expected Class is data.frame and type is list", {
  # Example 1 ---------------------------------------------------------------
  # skip if offline
  skip_if_offline()
  #
  expect_s3_class(
    newsboard_data(firstDate = '2023-06-24'), "data.frame")
  # A data frame is built from a list:
  expect_type(
    newsboard_data(firstDate = '2023-06-24'), "list")

  # Example 2 ---------------------------------------------------------------

  expect_s3_class(
    newsboard_data(firstDate = '2023-06-24', messageType = 'Mistrade'),
    "data.frame")
  # A data frame is built from a list:
  expect_type(
    newsboard_data(firstDate = '2023-06-24', messageType = 'Mistrade'),
    "list")


  # Example 3 ---------------------------------------------------------------
  expect_s3_class(
    newsboard_data(firstDate = '2023-06-24', Market = 'Six Swiss Exchange'),
    "data.frame")
  # A data frame is built from a list:
  expect_type(
    newsboard_data(firstDate = '2023-06-24', Market = 'Six Swiss Exchange'),
    "list")

  expect_s3_class(
    newsboard_data(firstDate = '2023-06-24', Products = 'Funds'),
    "data.frame")
  # A data frame is built from a list:
  expect_type(
    newsboard_data(firstDate = '2023-06-24', Products = 'Funds'),
    "list")

})
#> Test passed
#>

# test that the output from newsboard_data() returns a dataframe class object
test_that("Expected Error when firstDate < lastDate", {
  # Example 4 ---------------------------------------------------------------

  # skip if offline
  skip_if_offline()
  #
  # \dontrun{
  #data_final <- newsboard_data(lastDate = '2021-07-01') # error when lastdate > firstdate
  expect_error(
    newsboard_data(lastDate = '2021-07-01'),#data_final,
    fixed = TRUE,
    regexp = "SIX data ranges from firstDate to lastDate with firstDate < lastDate"
    #"SIX data ranges from firstDate to lastDate with firstDate < lastDate"
    )
  # }
  #
})
#> Test passed
#>
#>



R/logo_creation.R


# LOGO DATA ---------------------------------------------------------------
#
# swissexchangedata::newsboard_data(firstDate = '2023-10-10', lastDate = '2023-10-13') -> mm
# #
# mm_original <- mm
# # extract text and table from newsText through apply
# newsText_column_index <- grep("newsText", colnames(mm))
# #
# mm$data <- apply(
#   X = mm[,newsText_column_index,drop=FALSE],
#   MARGIN = 1, simplify = T,
#   FUN = itemlist_newsText_clean)
# #
# mm$table_data <- apply(
#   X = mm[,newsText_column_index,drop=FALSE],
#   MARGIN = 1, simplify = F,
#   FUN = itemlist_newsText_table_clean)
# #
# mm$table_data[[1]]
#
# logo_data <- rbind.data.frame(
#   (mm[[16]][[1]])[[1]][[1]]
#   ,(mm[[16]][[2]])[[1]][[1]]
#   ,(mm[[16]][[3]])[[1]][[1]]
#   ,(mm[[16]][[4]])[[1]][[1]]
# )
# logo_data
# #
# write.csv(logo_data, 'logo_data.csv', row.names = F)

#
#
# library(hexSticker)
# library(here)
# library(readr)
# library(ggplot2)
# library(gridExtra)
# # INtell CORP DATA from startdate to October 13th Adjusted closing prices
# INTC <- read_csv(
#   "INTC(2).csv",
#   col_types = cols(
#     Date = col_date(format = "%Y-%m-%d"),
#     Open  = col_skip(), High = col_skip(),
#     Low = col_skip(), Close  = col_skip(),
#     Volume = col_skip()))
# # logo data obtained from NEWBOARDS
# logo_data <- read_csv(
#   "logo_data.csv",
#   col_types = cols(
#     `Trade Date` = col_date(format = "%d.%m.%Y")))
# #View(logo_data)
# #
# subplot <- INTC |>
#   ggplot(aes(x = Date, y = `Adj Close`)) +
#   geom_area(fill = 'royalblue1') +
#   annotation_custom(
#     tableGrob(logo_data, rows = NULL,
#               theme = ttheme_minimal(
#                 base_size = 6,
#                 base_colour = "blue",
#                 parse = FALSE,
#                 padding = unit(c(2, 2), "mm"))),
#     xmin=6500, xmax=11500, ymin=35, ymax=60) +
#   theme_void()
# subplot
# #
# #
# # Create Sticker:
# sticker(subplot = subplot,                         # Subplot
#         s_x = 1,                            # Subplot x coordinate
#         s_y = 1.1,                         # Subplot y coordinate
#         s_width = 1.2,                      # Subplot width
#         s_height = 0.7,                       # Subplot height
#         package = "swissexchangedata", # Text
#         p_size = 16,                        # Text size
#         p_color = "#FFFFFF",                # Text color
#         p_y = 0.7,                         # Text y coordinate
#         h_fill = "#000000",                 # Hex background color
#         h_size = 3,                         # Hex border size
#         h_color = "#BB7351",                # Hex border color
#         dpi = 500,
#         filename = here(
#           "logo", "logo_swissexchangedata.png"))

#
# RDATA files in swissexchangedata\inst\ directory
# save(INTC, logo_data, file = ".\\inst\\extdata\\INTC_logo_data.RData")
#

#


R/newsboarddata.R
#' Products & Services > SIX Swiss Exchange Market Data > News & Tools > Newsboard data
#'
#' Clean scrapped data from the SIX Swiss Exchange Market Data website.
#'
#' @format ## `newsboardmarketdata`
#' A data frame with 1345 rows and 16 columns:
#' \describe{
#'   \item{messageNo}{Messae No}
#'   \item{isin}{Company, Symbol, ISIN}
#'   \item{valorSymbol}{Valor Symbol}
#'   \item{title}{Title}
#'   \item{messageType}{Message Type}
#'   \item{broadcastDateTime}{Broadcast Time}
#'   \item{security}{Security}
#'   \item{tradingSegment}{Trading Segment}
#'   \item{priority}{Priority}
#'   \item{markets}{Markets}
#'   \item{products}{Product}
#'   \item{currency}{Currency}
#'   \item{newsText}{News Text}
#'   \item{newsTypeCode}{News Type Code}
#'   \item{data}{Table Data}
#'   \item{table_data}{Cleaned and nested table data on trades}
#' }
#' @source <https://www.six-group.com/en/products-services/the-swiss-stock-exchange/market-data/news-tools/newsboard.html#/>
"newsboardmarketdata"


data-raw/newsboardmarketdata_raw_572021_572023.R
## code to cleaning and preparation `newsboardmarketdata_raw` dataset goes here

#
library(rvest)
#newsboardmarketdata_raw <- newsboard_data()
# obtain the preceding text
itemlist_newsText_clean <-function(vec){
  # obtain the preceding text
  read_html(vec) |> html_text()
}
# obtain the data table
itemlist_newsText_table_clean <-function(vec){

  # obtain the html tabke from the vector
  temp_obj = read_html(vec) |> html_table(fill=TRUE) #|>
  #  unlist() |> unlist()

  # if no data, skip, else convert to dataframe
    if (!is.null(temp_obj)){
      temp_obj = temp_obj |> as.matrix() |>
        t() |>
        as.data.frame()
    } else {
      NA
    }
  temp_obj
}

# get copy of data to newsboardmarketdata
newsboardmarketdata <- newsboardmarketdata_raw
# extract text and table from newsText through apply
newsText_column_index <- grep("newsText", colnames(newsboardmarketdata))
#
newsboardmarketdata$data <- apply(
  X = newsboardmarketdata[,newsText_column_index,drop=FALSE],
  MARGIN = 1, simplify = T,
  FUN = itemlist_newsText_clean)
#
newsboardmarketdata$table_data <- apply(
  X = newsboardmarketdata[,newsText_column_index,drop=FALSE],
  MARGIN = 1, simplify = T,
  FUN = itemlist_newsText_table_clean)
#
# newsboardmarketdata$table_data <- do.call(
#   rbind, lapply(
#     newsboardmarketdata$table_data, as.data.frame))
# save the newsboard data
usethis::use_data(newsboardmarketdata, overwrite = TRUE)
#


R/newsboard_data.R
#' The Swiss Stock Exchange Market Data News via the SIX Group Newsboard
#'
#' @description Returns the Swiss Stock Exchange Market Data News via the SIX Group Newsboard
#'
#' @param firstDate a character string indicating the year-month-date of From date.
#'      The default is equivalent to firstDate = `Sys.Date() - 365*2`, the earliest date for which
#'      data is available on the Newsboard.
#' @param lastDate a character string indicating the year-month-date of To date.
#'      The default is equivalent to `lastDate = Sys.Date()``, the last date in range for which
#'      data is available, but not greater than current date on the Newsboard.
#' @param messageType a vector of character strings of the message type.
#'      This must be one or combination of either of the strings "Emergency Information",
#'      "Mistrade", "Suspension", "Trading Information".The default is
#'      equivalent to messageType = 'All'.
#' @param Market a vector of character strings of the Market type.
#'      This must be one or combination of either of the strings "Six Swiss Exchange",
#'      "Structured Products","XBTR - Bilateral Trading Plattform". The default is
#'      equivalent to Market = 'All'.
#' @param Products a vector of character strings of the products type.
#'      This must be one or combination of either of the strings "Bonds", "Derivatives",
#'      "Equities", "Exchange Traded Products", "Funds".The default is
#'      equivalent to Products = 'All'.
#'
#' @details The inputs must all be character strings, with the date format as %Y-%m-%d
#'      (as determined by `as.Date()`). The 'messageType', 'Market', 'Products', if not
#'      supplied, all character string options are used, which leads to the
#'      messageType = 'All', Market = 'All' and Products = 'All' defaults.
#'      Supplying individual parameter requires the use of complete string
#'      as opposed to the use of abbreviations.
#'
#' @returns Swiss Stock Exchange Market Data News via the SIX Group Newsboard
#'
#' @author Akhapwoya Obuchere
#'
#' @importFrom plyr mapvalues
#' @importFrom jsonlite fromJSON
#'
#' @export
#'
#' @references Newsboard: Market activity at the Swiss Stock Exchange. SIX. https://www.six-group.com/en/products-services/the-swiss-stock-exchange/market-data/news-tools/newsboard.html#/
#'
#' @examples
#' library(plyr)
#' library(jsonlite)
#'
#' data_final <- newsboard_data(firstDate = '2023-06-24')
#' head(data_final, 1)
#' data_final <- newsboard_data(firstDate = '2023-06-24', messageType = 'Mistrade')
#' head(data_final, 2)
#' data_final <- newsboard_data(firstDate = '2023-06-24', Market = 'Six Swiss Exchange')
#' head(data_final,1)
#' data_final <- newsboard_data(firstDate = '2023-06-24', Products = 'Funds')
#' head(data_final, 1)
#' \dontrun{
#' data_final <- newsboard_data(lastDate = '2021-07-01') # error when lastdate > firstdate
#' head(data_final, 1)
#' }
#'
#'
newsboard_data <- function(
    firstDate = Sys.Date() - 365*2,
    lastDate = Sys.Date(),
    messageType = c(
      "Emergency Information", "Mistrade",
      "Suspension", "Trading Information"),
    Market = c(
      "Six Swiss Exchange", "Structured Products",
      "XBTR - Bilateral Trading Plattform"),
    Products = c(
      "Bonds", "Derivatives", "Equities",
      "Exchange Traded Products", "Funds")
){
  # Check dates
  if (!(as.Date(firstDate) && as.Date(lastDate))) {
    stop("x must be a character string in the form of a valid date %Y-%m-d%.")
  }
  #
  stopifnot( #
      "SIX data ranges from firstDate to lastDate with firstDate < lastDate"=firstDate <= lastDate # Threw out the Error as it is not captured when error testing with testthat
  )
  #
  if( firstDate < (Sys.Date() - 365*2)) {
    warning("SIX only provides historical data as far as 2 years from current Sys.Date(), defaulting to firstDate=Sys.Date().")
    lastDate <- (Sys.Date() - 365*2)
  }

  #'
  firstDate = gsub(pattern = '-', replacement = '', x = firstDate)
  lastDate = gsub(pattern = '-', replacement = '', x = lastDate)
  #'
  #' match
  if (!missing(messageType)){
    messageType = match.arg(
      arg = messageType,
      choices = c(
        "Emergency Information", "Mistrade",
        "Suspension", "Trading Information"),
      several.ok = TRUE)
    messageType <- plyr::mapvalues(
      messageType,
      from = c(
        "Emergency Information", "Mistrade",
        "Suspension", "Trading Information"),
      to = c("EI", "MI", "SU", "TI"), warn_missing = FALSE)
    messageType = paste0("messageType", messageType, '=true', sep = '', collapse = '&')
  } else {

    messageType = match.arg(
      arg = messageType,
      choices = c(
        "Emergency Information", "Mistrade",
        "Suspension", "Trading Information"),
      several.ok = TRUE)
    messageType <- mapvalues(
      messageType,
      from = c(
        "Emergency Information", "Mistrade",
        "Suspension", "Trading Information"),
      to = c("EI", "MI", "SU", "TI"), warn_missing = FALSE)
    #
    if (length(messageType) == 4){
      messageType = paste('messageTypeAll', '=true', sep = '')
    }
    #
  }
  ##############################################
  if (!missing(Market)){
    Market = match.arg(
      arg = Market,
      choices =  c(
        "Six Swiss Exchange", "Structured Products",
        "XBTR - Bilateral Trading Plattform"),
      several.ok = TRUE)
    Market <- plyr::mapvalues(
      Market,
      from = c(
        "Six Swiss Exchange", "Structured Products",
        "XBTR - Bilateral Trading Plattform"),
      to = c("XSWX", "XQMH", "XBTR"), warn_missing = FALSE)
    #
    Market = paste0("market", Market, '=true', sep = '', collapse = '&')
  } else {

    Market = match.arg(
      arg = Market,
      choices =  c(
        "Six Swiss Exchange", "Structured Products",
        "XBTR - Bilateral Trading Plattform"),
      several.ok = TRUE)
    Market <- plyr::mapvalues(
      Market,
      from = c(
        "Six Swiss Exchange", "Structured Products",
        "XBTR - Bilateral Trading Plattform"),
      to = c("XSWX", "XQMH", "XBTR"), warn_missing = FALSE)
    if (length(Market) == 3){
      Market = paste("marketAll", '=true', sep = '')
    }
    #
  }
  ########################################################
  if (!missing(Products)){
    Products = match.arg(
      arg = Products,
      choices = c(
        "Bonds", "Derivatives", "Equities",
        "Exchange Traded Products", "Funds"),
      several.ok = TRUE)

    Products <- plyr::mapvalues(
      Products,
      from = c(
        "Bonds", "Derivatives", "Equities", "Exchange Traded Products", "Funds"),
      to = c("BO","DE","EQ","EP", "FU"), warn_missing = FALSE)
    Products = paste0("products", Products, '=true', sep = '', collapse = '&')
  } else {
    Products = match.arg(
      arg = Products,
      choices = c(
        "Bonds", "Derivatives", "Equities",
        "Exchange Traded Products", "Funds"),
      several.ok = TRUE)

    Products <- plyr::mapvalues(
      Products,
      from = c(
        "Bonds", "Derivatives", "Equities", "Exchange Traded Products", "Funds"),
      to = c("BO","DE","EQ","EP", "FU"), warn_missing = FALSE)
    #
    if (length(Products) == 5){
      Products = paste("productTypeAll", '=true', sep = '')
    }
  }
  #
  url_root <-
    paste(
      "https://www.six-group.com/sheldon/newsboard/v1/find.json?firstDate=",
      firstDate, '&lastDate=', lastDate, "&pageSize=10&pageNumber=", sep = '')
  #    "https://www.six-group.com/sheldon/newsboard/v1/find.json?firstDate=firstDate&lastDate=lastDate&pageSize=10&pageNumber="
  #print(url)
  #url <- paste0(url, messageType, Market, Products, sep = '', collapse = '')
  #print(url)
  #print(lastDate)
  #
  trade_data_frame <- data.frame(NULL)
  #trade_details_data_frame <- data.frame(NULL)
  for (i in c(0:300)){ #Last value in seq based on size of number of results to last page
    # print page
    print(paste('Page', i))
    url <- paste(
      url_root, i, '&',
      paste(messageType, Market, Products, sep = '&'),
      sep = "")
    # print(url)
    # get data
    data <- jsonlite::fromJSON(
      #Captures the table of data within the page
      url
    )
    print(paste("Page",i, "has", nrow(data$itemList), "observations."))
    #stopifnot('END'= dim(data$itemList) !=  NULL) #Stops if next page is empty or rather past last page
    if (is.null(dim(data$itemList))){
      break
    }

    trade_data_frame = rbind(trade_data_frame, data$itemList)
    #
    Sys.sleep(0.0005)
    #
  }
  #
  # #Subset the desired Missedtrade and remove transactions with missing message numbers
  # trade_data_frame <- trade_data_frame[complete.cases(trade_data_frame$messageNo),]
  # string_value <- paste("trade Decision in CFRUSU / CH0589945366")
  # trade_data_frame <- subset(trade_data_frame, title == string_value)
  #
  trade_data_frame
  #
}
#'
#Sys.Date()
# data_final <- newsboard_data(firstDate = '2023-06-24',
#                               messageType = 'Mistrade')
#
# #
# data_final <- newsboard_data(firstDate = '2023-06-24',
#                              Market = 'XBTR - Bilateral Trading Plattform')
# #
# data_final <- newsboard_data(firstDate = '2023-06-24', Products = 'Funds')
#
#data_final <- newsboard_data()
#'
#'
#'
#'
# saveRDS(object = data_final, file = "R/data/newsboard_data.rds")
##
